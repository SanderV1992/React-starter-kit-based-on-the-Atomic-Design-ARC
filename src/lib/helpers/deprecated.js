let warned = {} // eslint-disable-line prefer-const

/**
 * @param message
 */
function warnOnce(message) {
  if (!warned[message]) {
    /* istanbul ignore else */
    if (typeof console !== 'undefined') {
      console.error(message) // eslint-disable-line no-console
    }
    warned[message] = true
  }
}

/**
 * Make propTypes property deprecated (Bootstrap realization)
 *
 * @param propType
 * @param explanation
 * @returns {validate}
 *
 * example: block:
 *  - toggleable: deprecated(PropTypes.bool, 'Please use the props "body"'),
 *  - toggleable: deprecated(
 *      PropTypes.oneOfType([PropTypes.bool, PropTypes.string]), 'Please use the prop "expand"'
 *  ),
 */
export function deprecated(propType, explanation) {
  return function validate(props, propName, componentName, ...rest) {
    if (props[propName] !== null && typeof props[propName] !== 'undefined') {
      warnOnce(`"${propName}" property of "${componentName}" has been deprecated.\n${explanation}`)
    }

    return propType(props, propName, componentName, ...rest)
  }
}
