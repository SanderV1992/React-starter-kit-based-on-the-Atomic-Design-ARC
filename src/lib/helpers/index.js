// helpers
export { em, rem } from './polished'
export { _ } from './lodash'
export { prop, ifProp, switchProp } from 'styled-tools'

// custom helpers
export { deprecated } from './deprecated'
export { palettePack as palette } from './palette'
