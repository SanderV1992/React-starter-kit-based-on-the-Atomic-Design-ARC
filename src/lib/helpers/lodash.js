import get from 'lodash/get'
import isNumber from 'lodash/isNumber'
import isString from 'lodash/isString'
import isArray from 'lodash/isArray'


/* eslint-disable id-match */
export const _ = {
  get,
  isNumber,
  isString,
  isArray,
}
