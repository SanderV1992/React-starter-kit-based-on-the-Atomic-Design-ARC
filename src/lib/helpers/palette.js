import get from 'lodash/get'


/* eslint-disable no-magic-numbers */
export const palettePack = {
  getAll: (packColor) => (props) => {
    const colorPack = props.theme.packs[packColor] || props.theme.packs[props.color]

    const { palette } = props.theme

    /* eslint-disable no-param-reassign */
    return Object.entries(colorPack).map(([key, value]) => {
      const [bgColor, bgShade] = value.background
      const [fgColor, fgShade] = value.foreground

      return {
        [key]: {
          background: palette[bgColor][bgShade || 500],
          foreground: palette[fgColor][fgShade || 500],
        },
      }
    }).reduce((obj, item) => ({ ...obj, ...item }), {})
  },
  getAllText: (packColor, textShade) => (props) => {
    const colorPack = props.theme.packs[packColor] || props.theme.packs[props.color]

    const { palette, paletteText } = props.theme

    return Object.entries(colorPack).map(([key, value]) => {
      const [bgColor, bgShade] = value.background
      const [fgColor, fgShade] = value.foreground

      return {
        [key]: {
          background: palette[paletteText[bgColor][bgShade || 500]][textShade || 'primary'],
          foreground: palette[paletteText[fgColor][fgShade || 500]][textShade || 'primary'],
        },
      }
    })
  },
  getColor: (path) => (props) => {
    const key = path.split('.')[0]
    const colorPack = props.theme.packs[key]
      ? {
        [key]: props.theme.packs[key],
      } : {
        [props.color]: props.theme.packs[props.color],
      }

    const newPath = props.theme.packs[key]
      ? path
      : `${props.color}.${path}`

    const { palette } = props.theme
    const [color, shade] = get(colorPack, newPath)

    return palette[color][shade || 500]
  },
  getColorText: (path, textShade) => (props) => {
    const key = path.split('.')[0]
    const colorPack = props.theme.packs[key]
      ? {
        [key]: props.theme.packs[key],
      }
      : {
        [props.color]: props.theme.packs[props.color],
      }

    const newPath = props.theme.packs[key]
      ? path
      : `${props.color}.${path}`

    const { palette, paletteText } = props.theme
    const [color, shade] = get(colorPack, newPath)

    return palette[paletteText[color][shade || 500]][textShade || 'primary']
  },
}
