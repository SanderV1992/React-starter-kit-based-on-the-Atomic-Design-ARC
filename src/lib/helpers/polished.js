import _rem from 'polished/lib/helpers/rem'
import { baseFontSize } from 'themes'


export em from 'polished/lib/helpers/em'

// Adapter
export const rem = (num) => _rem(num, baseFontSize)
