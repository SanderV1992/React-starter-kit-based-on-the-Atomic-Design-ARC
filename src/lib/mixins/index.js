export { fontFace } from './polished'
export { breakpoint } from './breakpoint'
