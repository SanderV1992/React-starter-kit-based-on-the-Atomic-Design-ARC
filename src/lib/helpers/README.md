# Utils

### Deprecated
```js
import { deprecated } from 'lib/helpers'

const stringOrNumberProp = PropTypes.oneOfType([PropTypes.number, PropTypes.string])

const columnProps = PropTypes.oneOfType([
  PropTypes.shape({
    size: PropTypes.oneOfType([PropTypes.bool, PropTypes.number, PropTypes.string]),
    push: deprecated(stringOrNumberProp, 'Please use the prop "order"'),
    pull: deprecated(stringOrNumberProp, 'Please use the prop "order"'),
    order: stringOrNumberProp,
    offset: stringOrNumberProp,
  })
])
```

### Breakpoint
```js
const button = css`
  ${breakpoint('mobile')} {
    background-color: #56c220;
  }
  
  ${breakpoint('tablet')} {
    background-color: #ccc;
  }
  
  ${breakpoint('desktop')} {
    background-color: #000;
  }
`
```

### Palette
```js
  color: ${palette('main.background')};
```
