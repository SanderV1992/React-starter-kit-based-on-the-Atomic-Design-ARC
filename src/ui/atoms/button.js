import React from 'react'
import PropTypes from 'prop-types'

import styled, { css } from 'styled-components'
import { palette, rem, prop, switchProp, _ } from 'lib/helpers'


const sizes = {
  xs: { height: 18, font: 9 },
  sm: { height: 28, font: 14 },
  md: { height: 38, font: 19 },
  lg: { height: 48, font: 24 },
}

export const Button = styled.button.attrs({
  fontSize: (p) => rem(sizes[p.size].font),
  height: (p) => rem(sizes[p.size].height),
  width: (p) => _.isNumber(p.width) ? rem(p.width) : p.width,
})`
  cursor: pointer;
  line-height: 0;
  outline: none;

  border-radius: ${prop('theme.button.borderRadius')};

  text-transform: ${prop('theme.button.textTransform')};
  font-weight: ${prop('theme.button.fontWeight')};
  font-family: ${prop('theme.button.fontFamily')};

  font-size: ${prop('fontSize')};
  height: ${prop('height')};
  width: ${prop('width')};

  /* @param {string} type - flat | raised */
  ${switchProp('type', {
    flat: css`
      color: ${palette.getColor('initial.background')};
      fill: ${palette.getColor('initial.background')};

      background-color: transparent;
      border: 1px solid ${palette.getColor('hover.background')};

      &:hover {
        background-color: ${palette.getColor('hover.foreground')};
        border-color: ${palette.getColor('hover.foreground')};
      }

      &:disabled, &:disabled:hover {
        cursor: default;
        color: ${palette.getColor('disabled.foreground')};
        fill: ${palette.getColor('disabled.foreground')};

        background-color: transparent;
        border-color: ${palette.getColor('disabled.foreground')};
      }
    `,
    raised: css`
      color: ${palette.getColorText('initial.background')};
      fill: ${palette.getColorText('initial.background')};

      background-color: ${palette.getColor('initial.background')};
      border: 1px solid ${palette.getColor('initial.background')};

      &:hover {
        background-color: ${palette.getColor('hover.background')};
        border-color: ${palette.getColor('hover.background')};
      }

      &:disabled, &:disabled:hover {
        cursor: default;
        color: ${palette.getColorText('disabled.background', 'disabled')};
        fill: ${palette.getColorText('disabled.background', 'disabled')};

        background-color: ${palette.getColor('disabled.background')};
        border-color: ${palette.getColor('disabled.background')};
      }
    `,
  })};
`

Button.defaultProps = {
  type: 'raised',
  color: 'default',
  size: 'md',
  width: 'auto',
  disabled: false,
}

Button.propTypes = {
  type: PropTypes.oneOf(['flat', 'raised']),
  color: PropTypes.oneOf(['default', 'primary', 'secondary', 'success', 'warning', 'danger']),
  size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg']),
  width: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]),
  disabled: PropTypes.bool,
}
