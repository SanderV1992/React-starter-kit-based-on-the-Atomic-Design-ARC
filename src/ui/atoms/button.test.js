import React from 'react'

import { ThemeProviderMock } from 'lib/helpers'
import { Button } from './button'
import renderer from 'react-test-renderer'


describe('Button', () => {
  test('Default Button (snapshot)', () => {
    const button = renderer.create(
      <ThemeProviderMock>
        <Button>Submit</Button>
      </ThemeProviderMock>
    ).toJSON()

    expect(button).toMatchSnapshot()
  })

  test('Button - size (snapshot)', () => {
    const tree = renderer.create(
      <ThemeProviderMock>
        <Button size="lg">Submit</Button>
      </ThemeProviderMock>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('Button - color (snapshot)', () => {
    const tree = renderer.create(
      <ThemeProviderMock>
        <Button color="primary">Submit</Button>
        <Button color="success">Submit</Button>
      </ThemeProviderMock>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('Button - disabled (snapshot)', () => {
    const tree = renderer.create(
      <ThemeProviderMock>
        <Button disabled>Submit</Button>
      </ThemeProviderMock>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('Button - all props (snapshot)', () => {
    const tree = renderer.create(
      <ThemeProviderMock>
        <Button color="success" size="lg" disabled>Submit</Button>
      </ThemeProviderMock>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
