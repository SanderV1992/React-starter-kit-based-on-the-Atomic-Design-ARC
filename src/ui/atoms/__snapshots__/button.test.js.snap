// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Button - all props (snapshot) 1`] = `
.c0 {
  outline: none;
  line-height: 0;
  cursor: pointer;
  -webkit-transition: opacity .2s linear;
  transition: opacity .2s linear;
  color: #fff;
  background-color: #28a745;
  border-color: #28a745;
  font-size: 1.5rem;
  height: 3rem;
  opacity: .6 !important;
}

.c0:hover {
  opacity: .8;
}

<button
  className="c0"
  disabled={true}
  onClick={[Function]}
>
  Submit
</button>
`;

exports[`Button Button - color (snapshot) 1`] = `
.c0 {
  outline: none;
  line-height: 0;
  cursor: pointer;
  -webkit-transition: opacity .2s linear;
  transition: opacity .2s linear;
  color: #fff;
  background-color: #28a745;
  border-color: #28a745;
  font-size: 1.1875rem;
  height: 2.375rem;
}

.c0:hover {
  opacity: .8;
}

<button
  className="c0"
  disabled={false}
  onClick={[Function]}
>
  Submit
</button>
`;

exports[`Button Button - disabled (snapshot) 1`] = `
.c0 {
  outline: none;
  line-height: 0;
  cursor: pointer;
  -webkit-transition: opacity .2s linear;
  transition: opacity .2s linear;
  color: #fff;
  background-color: #1890ff;
  border-color: #1890ff;
  font-size: 1.1875rem;
  height: 2.375rem;
  opacity: .6 !important;
}

.c0:hover {
  opacity: .8;
}

<button
  className="c0"
  disabled={true}
  onClick={[Function]}
>
  Submit
</button>
`;

exports[`Button Button - size (snapshot) 1`] = `
.c0 {
  outline: none;
  line-height: 0;
  cursor: pointer;
  -webkit-transition: opacity .2s linear;
  transition: opacity .2s linear;
  color: #fff;
  background-color: #1890ff;
  border-color: #1890ff;
  font-size: 1.5rem;
  height: 3rem;
}

.c0:hover {
  opacity: .8;
}

<button
  className="c0"
  disabled={false}
  onClick={[Function]}
>
  Submit
</button>
`;

exports[`Button Default Button (snapshot) 1`] = `
.c0 {
  outline: none;
  line-height: 0;
  cursor: pointer;
  -webkit-transition: opacity .2s linear;
  transition: opacity .2s linear;
  color: #fff;
  background-color: #1890ff;
  border-color: #1890ff;
  font-size: 1.1875rem;
  height: 2.375rem;
}

.c0:hover {
  opacity: .8;
}

<button
  className="c0"
  disabled={false}
  onClick={[Function]}
>
  Submit
</button>
`;
